README.md
# React Native Introduction Slider Component

A React Native based component for creating slider in your application. 

## Getting started

- [Demo](#demo)
- [Installation](#installation)
- [Supported versions](#supported-versions)
- [Usage](#usage)
- [Props](#props)
- [Methods](#methods)
- [Want to Contribute?](#want-to-contribute?)
- [Need Help?](#need-help?)
- [Collection of Components](#collection-of-components)
- [Changelog](#changelog)
- [License](#license)
- [Keywords](#keywords)

## Demo

[![](IntroductionSlider.png)](https://github.com/weblineindia/React-Native-Introduction-Slider/blob/master/IntroductionSlider.png)

## Installation

`$ npm install rn-weblineindia-intro-slider --save`

## Supported versions

We have tested this component in ReactNative 0.60 - 0.66.4. You can still use it in other versions.

## Usage

Create a new component file and regiser AppIntro component with props data

```javascript
import AppIntro from 'rn-weblineindia-intro-slider';

const propData = {
    sourceType: 'URL',
    navigateTo: 'SelectionScreen',
    defaultTint: 'gray',
    selectedTint: 'black',
    buttonStyle: {
      fontSize: 15,
      color: 'white',
      textAlign: 'right',
    },
    DATA: ["https://picsum.photos/200/501", "https://picsum.photos/300/502"],
  };

export default const IntroductionSlider = ({ navigation }) => {
    return <AppIntro 
      extraData={propData}
      navigation={navigation}
      onSkipTapped={() => { navigation.goBack() }}
    />
  }
 
```

Register IntroductionSlider Component to the navigation stack

```javascript
import IntroductionSlider from '../IntroductionSlider';

<Stack.Screen 
    name="IntroductionSlider" 
    component={IntroductionSlider}
/>
```

Here is a sample screen through which you can navigate to the IntroductionSlider Component

```javascript
<TouchableOpacity onPress={()=> navigation.navigate('IntroductionSlider')}>
    // Your code here
</TouchableOpacity>
```

### Callback

```javascript
this.props.onSkipTapped(this.state.index);

onSkipTapped={() => 
  // callback actions here...
}
```

## Props

| **Props**            | **Type** | **Required** | **Description**                                                               |
|----------------------|----------|--------------|-------------------------------------------------------------------------------|
| sourceType           | `string` | yes          | Using which source you want to load images, either assets or from url.        |
| navigateTo           | `string` | yes          | Route name you want to navigate when user tapped on Skip or Continue          |
| defaultTint          | `string` | no           | Default tint color for page-control                                           |
| selectedTint         | `string` | no           | Selected tint color for page-control                                          |
| buttonStyle          | `object` | no           | Style for Skip and Continue text                                              |
| DATA                 | `object` | yes          | Array of images, it contain assets or url                                     |
| isAutoScroll         | `boolean`| no           | default true                                                                  |
| timeDuration         | `number` | no           | default 5 sec, For better result use max 3 sec                                |

## Methods

| **Method**           | **Description**                                                             |
|----------------------|-----------------------------------------------------------------------------|
| onSkipTapped         | Callback method that is called when user pressed on Skip or Cancel button.  |

-----

## Want to Contribute?

- Created something awesome, made this code better, added some functionality, or whatever (this is the hardest part).
- [Fork it](http://help.github.com/forking/).
- Create new branch to contribute your changes.
- Commit all your changes to your branch.
- Submit a [pull request](http://help.github.com/pull-requests/).

-----

## Need Help? 

We also provide a free, basic support for all users who want to use this ReactNative Introduction Slider Component in their software project. In case you want to customize this Introduction Slider Component to suit your development needs, then feel free to contact our [ReactNative developers](https://www.weblineindia.com/hire-react-native-developers.html).

-----

## Collection of Components

We have built many other components and free resources for software development in various programming languages. Kindly click here to view our [Free Resources for Software Development](https://www.weblineindia.com/software-development-resources.html).

------

## Changelog

Detailed changes for each release are documented in [CHANGELOG.md](./CHANGELOG.md).

------

## License

[MIT](LICENSE)

[mit]: https://github.com/weblineindia/React-Native-Introduction-Slider/blob/master/LICENSE

------

### Keywords

 rn-weblineindia-intro-slider, react-native-intro-slider, intro-slider, introduction-slider, react-native


CHANGELOG.md
1.0.3 (Feb 03, 2022)
- Upgrade to Latest React Native version.

1.0.1 (July 30, 2020)
- We have integrated automatic slide mechanism.
- You can manage sliding duration using timeDuration props.
- You can manage manual sliding/auto sliding using isAutoScroll props.

1.0.0 (July 13, 2020)
Initial release


styles.js
import { StyleSheet, Dimensions } from 'react-native';

const ScreenWidth = Dimensions.get('window').width;
const ScreenHeight = Dimensions.get('window').height;

const styleSheet = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: 'transparent',
  },
  flatList: {
    flex: 1,
    // backgroundColor: 'gray',
  },
  footer: {
    position: 'absolute',
    height: '7%',
    width: ScreenWidth,
    left: 0,
    bottom: 0,
    // flexDirection: 'row',
    // alignSelf: 'center',
    // alignItems: 'center',
    // justifyContent: 'center',
    // backgroundColor: 'purple',
  },
  pageControl: {
    flex: 3,
    height: '100%',
    // backgroundColor: 'green',
    alignItems: 'center',
    // marginBottom: ScreenWidth * 0.15,
    // alignSelf: 'flex-start',
  },
  skipSlideShow: {
    width: '100%',
    height: 30,
  },
  initAgreeButton: {
    width: ScreenWidth * 0.3,
    // height: ScreenWidth * 0.13,
    // backgroundColor: 'green',
    alignSelf: 'flex-end',
    justifyContent: 'center',
    // alignItems: 'center',
    fontSize: 13,
    borderRadius: ScreenHeight * 0.5,
    marginRight: ScreenWidth * 0.03,
  },
  initAgreeBtnText: {
    fontSize: 15,
    color: '#FFF',
    textAlign: 'right',
    // backgroundColor: 'blue',
    // fontFamily: RobotoMedium,
  },

  // --- SlideShow Item - CSS ---
  bannerView: {
    flex: 1,
  },
  banner: {
    height: ScreenHeight,
    width: ScreenWidth,
  },
  // --- PageControl Item - CSS ---
  dotListView: {
    flex: 1,
    justifyContent: 'center',
  },
  dotView: {
    width: 7,
    height: 7,
    borderRadius: 5,
    // backgroundColor: 'green',
    marginHorizontal: 4,
  },
});
export default styleSheet;


android/src/main/java/com/reactlibrary/RnWeblineindiaIntroSliderModule.java
package com.reactlibrary;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.Callback;

public class RnWeblineindiaIntroSliderModule extends ReactContextBaseJavaModule {

    private final ReactApplicationContext reactContext;

    public RnWeblineindiaIntroSliderModule(ReactApplicationContext reactContext) {
        super(reactContext);
        this.reactContext = reactContext;
    }

    @Override
    public String getName() {
        return "RnWeblineindiaIntroSlider";
    }

    @ReactMethod
    public void sampleMethod(String stringArgument, int numberArgument, Callback callback) {
        // TODO: Implement some actually useful functionality
        callback.invoke("Received numberArgument: " + numberArgument + " stringArgument: " + stringArgument);
    }
}


index.js
/**
Author Name  :  WeblineIndia  |  https://www.weblineindia.com/

Support : In case you want to customize this code library to suit your development needs, then feel free to contact.

For more such software development components and code libraries, visit us at
https://www.weblineindia.com/software-development-resources.html 

Our Github URL : https://github.com/weblineindia
**/
/* eslint-disable react/no-access-state-in-setstate */
import React from 'react';
import {
  View, TouchableOpacity, FlatList, ImageBackground, Text, StatusBar, Alert,
} from 'react-native';
import styleSheet1 from './styles';


type IProps = {
  navigation: Object,
  setAppIntroData: Function,
  extraData: Object,
  defaultTint: String,
  selectedTint: String,
  onSkipTapped: Function,
};

type IState = {
  pageIndex: Number,
  isFromPageControl: Boolean,
};

let timeInterval;
var animaInterval = null;

export default class AppIntro extends React.Component<IProps, IState> {
  constructor(props) {
    super(props);
    this.state = {
      pageIndex: 0,
      isFromPageControl: true,
    };
    console.log('--props--', this.props.extraData.defaultTint);
    if (this.props.extraData.defaultTint === undefined) {
      this.props.extraData.defaultTint = 'gray';
    }
    if (this.props.extraData.selectedTint === undefined) {
      this.props.extraData.selectedTint = 'lightgray';
    }
    if (this.props.extraData.buttonStyle === undefined) {
      this.props.extraData.buttonStyle = {
        fontSize: 15,
        color: '#000000',
        textAlign: 'right',
      };
    }
    if (this.props.extraData.isAutoScroll === true) {
      if(this.props.extraData.timeDuration === undefined){
        timeInterval = 3000;
      } else if (this.props.extraData.timeDuration <= 2000){
        timeInterval = 3000;
      } else {
        timeInterval = this.props.extraData.timeDuration;
      }
      this.setIntervalToScroll();
    } else {
      clearInterval(animaInterval);
    }
  }

  setIntervalToScroll() {


    // To set time interval for jump to next page

    animaInterval = setInterval(
      function () {
        if (this.props.extraData.isAutoScroll === true) {
          const { pageIndex } = this.state;
          let nextIndex = 0;

          if (pageIndex < this.props.extraData.DATA.length - 1) {
            nextIndex = pageIndex + 1;
          }

          this.setState({ pageIndex: nextIndex }, () => {
            if (nextIndex === (this.props.extraData.DATA.length - 1)){
              this.scrollToIndex(nextIndex, nextIndex === 0 ? false : true);
              clearInterval(animaInterval);
            } else {
              this.scrollToIndex(nextIndex, nextIndex === 0 ? false : true);
            }
          });
        } else {
          clearInterval(animaInterval);
        }
      }.bind(this),
      timeInterval,
    );
  }

  /**
   * Method for verifying list refrence exists or not
   * Apply animation when move to next page or index with timeout
   *
   * @param {*} index
   * @param {*} animated
   */

  scrollToIndex = (index, animated) => {
    this.ref_flatList && this.ref_flatList.scrollToIndex({ index, animated });
    this.flatList_Ref && this.flatList_Ref.scrollToIndex({ index, animated });
  };

  onScrollEnd(e) {
    const contentOffset = e.nativeEvent.contentOffset;
    const viewSize = e.nativeEvent.layoutMeasurement;

    // Divide the horizontal offset by the width of the view to see which page is visible
    const pageNum = Math.round(contentOffset.x / viewSize.width);
    if (this.state.isFromPageControl) {
      this.setState({ pageIndex: pageNum });
    }
    console.log('--pageNum--', pageNum, viewSize.width);
    setTimeout(() => {
      this.flatList_Ref.scrollToIndex({ animated: true, index: pageNum });
    }, 0.5);
  }

  renderBanner = (banner) => (
    <View style={styleSheet1.bannerView}>
      <ImageBackground
        style={styleSheet1.banner}
        source={this.props.extraData.sourceType === 'LOCAL' ? banner.bannerImage : { uri: banner }}
        resizeMode="cover"
      />
    </View>
  );

  renderPageControl = (index) => (
    <View style={styleSheet1.dotContainer}>
      <View
        style={[styleSheet1.dotView,
          {
            backgroundColor: this.state.pageIndex === index ?
              this.props.extraData.selectedTint : this.props.extraData.defaultTint,
          },
        ]}
      />
    </View>
  );

  closeHelp() {
    if (this.props.onSkipTapped) {
      this.props.onSkipTapped(this.state.index);
    }
    clearInterval(animaInterval);
    this.props.navigation.navigate('SelectionScreen');
  }

  scrollNext() {
    if (this.state.pageIndex < this.props.extraData.DATA.length - 1) {
      this.setState({ isFromPageControl: false, pageIndex: this.state.pageIndex + 1 }, () => {
        this.ref_flatList.scrollToIndex({ animated: true, index: this.state.pageIndex });
      });
    }
  }

  scrollPrev() {
    if (this.state.pageIndex > 0) {
      this.setState({ isFromPageControl: false, pageIndex: this.state.pageIndex - 1 }, () => {
        this.ref_flatList.scrollToIndex({ animated: true, index: this.state.pageIndex });
      });
    }
  }

  render() {
    console.log('render pageNum - ', this.state.pageIndex, this.state.pageIndex);
    return (
      <View style={styleSheet1.container}>
        <StatusBar translucent backgroundColor="transparent" barStyle="dark-content" />
        <FlatList // first
          horizontal
          pagingEnabled
          bounces={false}
          showsHorizontalScrollIndicator={false}
          legacyImplementation={false}
          style={styleSheet1.flatList}
          data={this.props.extraData.DATA}
          renderItem={({ item, index }) => this.renderBanner(item, index)}
          keyExtractor={(banner) => banner.id}
          onMomentumScrollEnd={(e) => this.onScrollEnd(e)}
          onScrollBeginDrag={() => this.setState({ isFromPageControl: true })}
          ref={(refe) => {
            this.ref_flatList = refe;
          }}
        />
        <View style={styleSheet1.footer}>
          <View style={styleSheet1.pageControl}>
            <FlatList // second
              scrollEnabled={false}
              horizontal
              bounces={false}
              showsHorizontalScrollIndicator={false}
              style={styleSheet1.flatList}
              data={this.props.extraData.DATA}
              renderItem={({ index }) => this.renderPageControl(index)}
              keyExtractor={(banner) => banner.id}
              ref={(ref) => {
                this.flatList_Ref = ref;
              }}
            />
          </View>

          <View style={styleSheet1.skipSlideShow}>
            <TouchableOpacity onPress={() => this.closeHelp()}>
              <View style={styleSheet1.initAgreeButton}>
                <Text style={this.props.extraData.buttonStyle}>{this.state.pageIndex === (this.props.extraData.DATA.length - 1) ? 'Continue' : 'Skip'}</Text>
              </View>
            </TouchableOpacity>
          </View>
        </View>

      </View>
    );
  }
}

AppIntro.defaultProps = {
  extraData: {
    defaultTint: 'white',
    selectedTint: 'darkgray',
  },
};
